### ISSUE TITLE
var.get()->requires_grad idiom is unchecked and allows silent memory corruption

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#4215]
Steps to reproduce:

```
void f(Variable& var) {
  var.get()->grad_fn = nullptr;
}
int main() {
  Variable v;
  f(v);
}
```

Expected result: runtime error of some sort.

Actual result: segmentation fault on program shutdown

You can try it out yourself by applying the following patch:

```
diff --git a/torch/csrc/jit/test_jit.cpp b/torch/csrc/jit/test_jit.cpp
index 0087310..12ccb64 100644
--- a/torch/csrc/jit/test_jit.cpp
+++ b/torch/csrc/jit/test_jit.cpp
@@ -8,6 +8,7 @@
 #include "torch/csrc/jit/interned_strings.h"
 #include <vector>
 #include "torch/csrc/jit/interpreter.h"
+#include "torch/csrc/autograd/variable.h"
 
 namespace torch { namespace jit {
 
@@ -496,7 +497,14 @@ void interpStageTest() {
     JIT_ASSERT(exactlyEqual(outputs[1],cx));
 }
 
+void f(torch::autograd::Variable& var) {
+  var.get()->_grad_fn = nullptr;
+}
+
 void runJITCPPTests() {
+  torch::autograd::Variable v;
+  f(v);
+  return;
   interpTest();
   interpStageTest();
   codeTemplateTest();
```

and then running `python test/test_jit.py TestJit.test_cpp`

I boiled this test case down while chasing down a very difficult to reproduce memory corruption.

cc @suo @colesbury

**Comment by apaszke** [comment]
I'd say that messing with `*Impl` is clearly not recommended, as they aren't part of the public API.

**Comment by ezyang** [comment]
OK, at the time I filed this, I wasn't quite sure if this was causing my memory corruption, but I've run some more tests and I am fairly confident that the memory corruption that I've been tracking down since Friday boils down to this function implementation in `VariableType.cpp`:

```
static void set_flags(at::ArrayRef<Variable> vl, VarFlags flags, std::shared_ptr<Function> grad_fn) {
  if (grad_fn) {
    grad_fn->num_inputs = vl.size();
  }
  int64_t output_nr = 0;
  for (auto& var : vl) {
    // TODO: combine this with the Variable construction
    var.get()->requires_grad = flags.requires_grad;
    var.get()->is_volatile = flags.is_volatile;
    var.get()->output_nr = output_nr;
    var.get()->_grad_fn = grad_fn;
    output_nr++;
  }
}
```

This code looked totally reasonable in review, and is in actuality extremely dangerous because it assumes that all the Variables are defined and silently corrupts memory if they are not. But *unlike* null pointers, the memory `var` points to is usually mapped, which means you don't get an immediate segfault when this happens.

Under what circumstances could we return an undefined variable from a method wrapped in Variable? The situation that triggers this is when we have backwards functions, which take an `output_mask` to determine what outputs they need to compute. Here, an undefined is returned in any case the gradient was not requested. Thus, making Variable non-default constructible doesn't solve the problem.

What about making `set_flags` a method on `Variable`, and truly making it illegal, everywhere, to access data members directly? I think that would make it sufficiently difficult to commit the same error in the future.

**Comment by ezyang** [comment]
@colesbury suggests padding out UndefinedTensorType with enough space so that writes don't corrupt other things. That probably would help avoid the segfault, although it would mask this kind of bug.

**Comment by apaszke** [comment]
Ugh, that doesn’t seem like a very safe thing. You’re basically ignoring undefined behavior in your program (I think of undefined tensors like null pointers). It’s like mapping memory at address 0 to prevent segfaults :smile:

**Comment by gchanan** [comment]
I'm not a huge fan of the memory mapping either.  The first thing that seems wrong to me is that the pImpl is not actually a `VariableImpl`, i.e. this definition is just wrong:
```
inline VariableImpl* Variable::get() const {
  return static_cast<VariableImpl*>(pImpl);
}
```
Why don't we wrap the undefined tensor in a VariableImpl when we create the Variable?  That should solve the memory corruption issue without playing any memory games.

It's a little weird to be able to set a grad_fn on a Variable that holds an undefined tensor, but we can decide if we want to provide functions that check that sort of thing.

**Comment by colesbury** [comment]
That would require allocation for every empty tensor, which we don't want. The original code is buggy. You shouldn't be calling anything on undefined tensors except `.defined()`. The question is simply do you want to mitigate the bad behavior, and can we do it without impacting performance.

**Comment by colesbury** [comment]
By far, the easiest way to detect these bugs is to use `nullptr` instead of `UndefinedTensor`. You'll get a segfault immediately instead of memory corruption or spurious reads.

**Comment by zdevito** [comment]
FYI, I hit the same problems here with interpreter code -- it is very is to be given a Variable object and assume if it is undefined that .data() would give an undefined Tensor, but instead it silently returns a nullptr which manifests later.

**Comment by ezyang** [comment]
I am implementing the mmap trick to help me debug another memory corruption which I suspect is due to an undefined tensor. I'm putting up the basic nuts and bolts of it up here in case people want to give implementation feedback:

```
namespace {

  constexpr size_t undefined_region_size = 4096; // one page

  UndefinedType* newUndefinedType(Context * context) {
    static_assert(sizeof(UndefinedType) <= undefined_region_size, "UndefinedType too large");
    void *p = mmap(NULL, undefined_region_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
    new (p) UndefinedType(context);
    mprotect(p, undefined_region_size, PROT_READ);
    return static_cast<UndefinedType*>(p);
  }

  void deleteUndefinedType(UndefinedType *p) {
    if (reinterpret_cast<uintptr_t>(static_cast<void*>(p)) % undefined_region_size != 0) {
      throw std::runtime_error("unaligned UndefinedType pointer, probably garbage\n");
    }
    p->~UndefinedType();
    mprotect(p, undefined_region_size, PROT_READ | PROT_WRITE);
    munmap(p, undefined_region_size);
  }

} // anonymous namespace

void Type::registerAll(Context * context) {
  ${type_registrations}
  context->type_registry[static_cast<int>(Backend::Undefined)][static_cast<int>(ScalarType::Undefined)] = { newUndefinedType(context), [](Type *x) { deleteUndefinedType(static_cast<UndefinedType*>(x)); } };
}
```

**Comment by eellison** [comment]
has there been any resolution on this ? I'm not sure this is a JIT issue.

**Comment by ezyang** [comment]
Since we merged variable and tensor in C++ this is no longer a problem!
